### DISCUSSION TITLE
`pnpm install` suddenly starting to fail with ERR_PNPM_ADDING_TO_ROOT error

### DISCUSSION THREAD

**Discussion by alancoppin** [vercel/community#3467] [Help]
### Summary

Hi everyone,

I'm using Remix and use `pnpm install` as the vercel install command, the same deployment that worked 5 days ago is now failing with the below error. Any ideas on how to resolve it?
I have tried adding `-w` to specify the root as the workspace but still an issue. Locally I don't have any issue running `pnpm i`

![image](https://github.com/vercel/community/assets/22746959/20bbdde2-d382-4020-bb57-f2b5c2e22b47)

### Deployment URL
https://connect-convert-remix-3foyfz9u0-connect-convert.vercel.app

**Comment by TooTallNate** [comment]
Can you share the deployment URL?

**Reply by alancoppin** [reply]
https://connect-convert-remix-3foyfz9u0-connect-convert.vercel.app

**Reply by TooTallNate** [reply]
Ok, I think I have a fix (https://github.com/vercel/vercel/pull/10291). Can you try with this environment variable set to confirm?

* Name: `VERCEL_CLI_VERSION`
* Value: `https://vercel-jtry5m27f.vercel.sh/tarballs/vercel.tgz`

**Reply by alancoppin** [reply]
@TooTallNate thank you! It's working now

**Comment by mmachatschek** [comment]
This may be related but we see the error in our monorepo setup with remix apps, but not with nextjs apps etc.:

ERR_PNPM_INCLUDED_DEPS_CONFLICT

![image](https://github.com/vercel/community/assets/10237069/93a060c8-5a5f-41c7-adec-bd9e346b14d5)

This happens since vercel cli version 31.2.0 and still happens with 31.2.2

**Reply by TooTallNate** [reply]
Can you share your `package.json` files (root level and for your Remix app)?

**Reply by mmachatschek** [reply]
@TooTallNate I'm afraid I can't share the files here publicly on GH but here is a deployment ID `4sPvZbmRTqzaam4wf6MFHXKeiGrW` if this is of any help.

I try to set up a reproducible build if you are unable to view that deployment

**Reply by zkochan** [reply]
Do you use the `optional`, `dev`, or `prod` settings? This may happen when you change this setting between to runs of `pnpm install`, `pnpm add`, or `pnpm update`.

**Reply by mmachatschek** [reply]
@zkochan the install command for our remix app is:

We need to use NODE_ENV development because we cant get rid of some dependencies in our dev deps for production builds.

The install commands didn't change. Just the vercel cli version

I hope I understood your question correctly

```
NODE_ENV=development pnpm install --filter=xyz...
```

**Reply by mmachatschek** [reply]
The only reference I see to this issue here is the PR that has been merged and release with the 31.2.0 release of the vercel cli https://github.com/vercel/vercel/pull/9784

As mentioned above, our CI on Vercel started to break without any changes to dependencies and without changes to the install commands

**Reply by TooTallNate** [reply]
It is definitely related to that PR. As a temporary measure, you can pin to an older CLI version, but setting env var `VERCEL_CLI_VERSION` with a value of `vercel@31.1.1`.

**Reply by k9ordon** [reply]
I just tested it out with `vercel@31.4.0` and it seems the issue persists.

**Reply by mmachatschek** [reply]
Issue still present in latest version `vercel@32.1.0`

**Reply by mmachatschek** [reply]
@TooTallNate after some digging, I found out what the culprit is/was

In our Vercel environment variables we had `NODE_ENV=production` set by default. In order to use the dev tools in our build step, we added `NODE_ENV=development` to the install command (as mentioned here: https://github.com/orgs/vercel/discussions/3467#discussioncomment-6666507).

With the update from the PR https://github.com/vercel/vercel/pull/9784, the vercel cli now tries to link `@remix-run/dev@npm:@vercel/remix-run-dev@1.x.x`.
As we set a custom `NODE_ENV` in our install command, the auto generated `pnpm install` command https://github.com/vercel/vercel/pull/9784/files#diff-15932333efa1bc7d772431f5ad740ced2793abf9d617280a73d5cd582a02649dR200-R205 by the vercel cli will not know how the node_modules have been installed.
In the workspace setup, PNPM will detect that we initially installed the `node_modules` with `NODE_ENV=development` but in the second step (executed by vercel cli) it falls back to the original environment variables set in the vercel dashboard (which is `NODE_ENV=production`) which PNPM doesn't like as we have a mismatch in the installed dependencies https://github.com/pnpm/pnpm/blob/46cecfc60fcb38f445df05ab160dd552282c7426/pkg-manager/get-context/src/index.ts#L272-L283.

So for us, the workaround/fix is to delete the `NODE_ENV=production` from the environment variables in the vercel dashboard, and inject it in the build command. Therefore the `installCommand` of our script will have the same `NODE_ENV` as the vercel cli

The issue I have with the current implementation of the [remix specific code](https://github.com/vercel/vercel/blob/60c75fd76cf45a53b8143c560e295a2aed62b450/packages/remix/src/build.ts#L208-L212) on the vercel cli is, that it ignores possible flags that have been set on the `installCommand`.

e.g. the vercel cli remix integration will ignore `PRISMA_SKIP_POSTINSTALL_GENERATE=true` set in the `installCommand` like so: `PRISMA_SKIP_POSTINSTALL_GENERATE=true pnpm install blog...` this means that the second install of the updated/added packages by vercel cli will not add that flag and potentially execute postinstall/lifecycle scripts that should/shouldn't be executed.
